File: react_on_rails_pro/lib/react_on_rails_pro/license_refresh_checker.rb
# frozen_string_literal: true

module ReactOnRailsPro
  # Handles the timing logic for automatic license refresh.
  # Determines when to check for a refreshed license based on expiry proximity.
  #
  # Refresh schedule:
  # - <= 7 days until expiry: check daily
  # - <= 30 days until expiry: check weekly
  # - > 30 days: no refresh needed
  module LicenseRefreshChecker
    class << self
      def maybe_refresh_license
        return unless ReactOnRailsPro.configuration.auto_refresh_enabled?
        return unless should_check_for_refresh?

        response = LicenseFetcher.fetch
        return if response.nil?

        LicenseCache.write(response)
      end

      # Seeds the cache on first boot so that refresh logic works on subsequent boots.
      # The cache stores the token's expiry, which should_check_for_refresh? needs to determine
      # when to trigger a refresh.
      def seed_cache_if_needed(license_data)
        return unless ReactOnRailsPro.configuration.auto_refresh_enabled?
        return if LicenseCache.token.present? # Cache already exists

        token = load_token_from_env_or_file
        return unless token

        expires_at = Time.at(license_data["exp"])

        LicenseCache.write(
          "token" => token,
          "expires_at" => expires_at.iso8601
        )
      end

      def should_check_for_refresh?
        expires_at = LicenseCache.expires_at
        return false if expires_at.nil?

        days_until_expiry = ((expires_at - Time.now) / 1.day).to_i

        if days_until_expiry <= 7
          last_fetch_older_than?(1.day)
        elsif days_until_expiry <= 30
          last_fetch_older_than?(7.days)
        else
          false
        end
      end

      def last_fetch_older_than?(duration)
        fetched_at = LicenseCache.fetched_at
        return true if fetched_at.nil?

        Time.now - fetched_at > duration
      end

      # Loads token from ENV or file, skipping cache (used for seeding)
      def load_token_from_env_or_file
        license = ENV.fetch("REACT_ON_RAILS_PRO_LICENSE", nil)
        return license if license.present?

        config_path = Rails.root.join("config", "react_on_rails_pro_license.key")
        return File.read(config_path).strip if config_path.exist?

        nil
      end
    end
  end
end

File: packages/react-on-rails-pro-node-renderer/src/shared/licenseRefreshChecker.ts
/**
 * Handles the timing logic for automatic license refresh.
 * Determines when to check for a refreshed license based on expiry proximity.
 *
 * Refresh schedule:
 * - <= 7 days until expiry: check daily
 * - <= 30 days until expiry: check weekly
 * - > 30 days: no refresh needed
 *
 * @module shared/licenseRefreshChecker
 */

import * as fs from 'fs';
import * as path from 'path';
import { isAutoRefreshEnabled, fetchLicense } from './licenseFetcher.js';
import { getCachedToken, getFetchedAt, getExpiresAt, writeCache } from './licenseCache.js';

const ONE_DAY_MS = 24 * 60 * 60 * 1000;
const SEVEN_DAYS_MS = 7 * ONE_DAY_MS;

/**
 * Checks if the last fetch was older than the specified duration.
 */
function lastFetchOlderThan(durationMs: number): boolean {
  const fetchedAt = getFetchedAt();
  if (!fetchedAt) {
    return true;
  }

  return Date.now() - fetchedAt.getTime() > durationMs;
}

/**
 * Determines if we should check for a license refresh based on expiry timing.
 */
export function shouldCheckForRefresh(): boolean {
  const expiresAt = getExpiresAt();
  if (!expiresAt) {
    return false;
  }

  const daysUntilExpiry = Math.floor((expiresAt.getTime() - Date.now()) / ONE_DAY_MS);

  if (daysUntilExpiry <= 7) {
    // Check daily when within 7 days of expiry
    return lastFetchOlderThan(ONE_DAY_MS);
  }
  if (daysUntilExpiry <= 30) {
    // Check weekly when within 30 days of expiry
    return lastFetchOlderThan(SEVEN_DAYS_MS);
  }

  // No refresh needed when more than 30 days until expiry
  return false;
}

/**
 * Attempts to refresh the license if conditions are met.
 * This is an async operation that fetches from the API and updates the cache.
 */
export async function maybeRefreshLicense(): Promise<void> {
  if (!isAutoRefreshEnabled()) {
    return;
  }

  if (!shouldCheckForRefresh()) {
    return;
  }

  const response = await fetchLicense();
  if (!response) {
    return;
  }

  writeCache({
    token: response.token,
    expires_at: response.expires_at,
  });
}

/**
 * Loads the license token from ENV or config file (not cache).
 * Used for seeding the cache on first boot.
 */
function loadTokenFromEnvOrFile(): string | null {
  // Check environment variable first
  const envLicense = process.env.REACT_ON_RAILS_PRO_LICENSE;
  if (envLicense) {
    return envLicense;
  }

  // Try config file
  try {
    const configPath = path.join(process.cwd(), 'config', 'react_on_rails_pro_license.key');
    if (fs.existsSync(configPath)) {
      return fs.readFileSync(configPath, 'utf8').trim();
    }
  } catch {
    // Ignore errors reading config file
  }

  return null;
}

/**
 * Seeds the cache on first boot so that refresh logic works on subsequent boots.
 * The cache stores the token's expiry, which shouldCheckForRefresh() needs to determine
 * when to trigger a refresh.
 *
 * @param licenseExp - The expiration timestamp from the validated license
 */
export function seedCacheIfNeeded(licenseExp: number): void {
  if (!isAutoRefreshEnabled()) {
    return;
  }

  // Cache already exists
  if (getCachedToken()) {
    return;
  }

  const token = loadTokenFromEnvOrFile();
  if (!token) {
    return;
  }

  const expiresAt = new Date(licenseExp * 1000);

  writeCache({
    token,
    expires_at: expiresAt.toISOString(),
  });
}
